addons/mail/static/src/models/message_action.js

```diff
@@ -0,0 +1,148 @@
+/** @odoo-module **/
+
+import { registerModel } from '@mail/model/model_core';
+import { attr, one } from '@mail/model/model_field';
+import { clear, insertAndReplace } from '@mail/model/model_field_command';
+
+registerModel({
+    name: 'MessageAction',
+    identifyingMode: 'xor',
+    recordMethods: {
+        /**
+         * @private
+         * @returns {integer}
+         */
+        _computeIsNonCompactActionContribution() {
+            if (this.messageActionListOwnerAsToggleCompact) {
+                return 0;
+            }
+            return 1;
+        },
+        /**
+         * @private
+         * @returns {FieldCommand}
+         */
+        _computeMessageActionListOwner() {
+            if (this.messageActionListOwnerAsDelete) {
+                return this.messageActionListOwnerAsDelete;
+            } else if (this.messageActionListOwnerAsEdit) {
+                return this.messageActionListOwnerAsEdit;
+            } else if (this.messageActionListOwnerAsMarkAsRead) {
+                return this.messageActionListOwnerAsMarkAsRead;
+            } else if (this.messageActionListOwnerAsReaction) {
+                return this.messageActionListOwnerAsReaction;
+            } else if (this.messageActionListOwnerAsReplyTo) {
+                return this.messageActionListOwnerAsReplyTo;
+            } else if (this.messageActionListOwnerAsToggleCompact) {
+                return this.messageActionListOwnerAsToggleCompact;
+            } else if (this.messageActionListOwnerAsToggleStar) {
+                return this.messageActionListOwnerAsToggleStar;
+            }
+            return clear();
+        },
+        /**
+         * @private
+         * @returns {FieldCommand}
+         */
+        _computeMessageActionView() {
+            /**
+             * Case 0: Always display Reaction and ToggleCompact if they are excisting.
+             * Case 1: If ToggleCompact doesn't excist in the action list, display the action.
+             * Case 2: If ToggleCompact excists in the action list and
+             *         the action list is not in the compact mode, display the action.
+             * Others: Do not display the action in the action list.
+             */
+            if (
+                this.messageActionListOwnerAsReaction ||
+                this.messageActionListOwnerAsToggleCompact ||
+                (
+                    this.messageActionListOwner &&
+                    !this.messageActionListOwner.actionToggleCompact ||
+                    (
+                        this.messageActionListOwner.actionToggleCompact &&
+                        !this.messageActionListOwner.isCompact
+                    )
+                )
```

Comment:
```suggestion
                this.messageActionListOwnerAsReaction ||
                this.messageActionListOwnerAsToggleCompact ||
                (this.messageActionListOwner && !this.messageActionListOwner.actionToggleCompact) ||
                (this.messageActionListOwner.actionToggleCompact && !this.messageActionListOwner.isCompact)
```

And why is there a guard on `messageActionListOwner` on the 3rd line and not the 4th?
Either the one on the 3rd is not useful, or the one on the 4th is missing.

Reply:
cause if the `this.messageActionListOwner` is false, the rest of the code will be skipped. So, it's basically it's a guard for all the code after that and we don't have to redundant the code.

if I change the code according to your suggestion, the `messageActionListOwner` should be added to the 4th line as well. I will do that.